nixos/modules/services/databases/victoriametrics/default.nix

```diff
@@ -0,0 +1,186 @@
+{
+  config,
+  pkgs,
+  lib,
+  ...
+}:
+with lib;
+let
+  cfg = config.services.victoriametrics;
+  yaml = pkgs.formats.yaml { };
+
+  promTypes = import ./promTypes.nix { inherit lib; };
+
+  bindAddr = "${cfg.listenAddress}:${builtins.toString cfg.port}";
+  workingDir = "/var/lib/" + cfg.stateDir;
+
+  generatedPrometheusYml = yaml.generate "prometheus.yml" scrapeConfig;
+
+  # This becomes the main config file for VictoriaMetrics's `-promscrape.config`
+  # https://docs.victoriametrics.com/vmagent/#how-to-collect-metrics-in-prometheus-format
+  scrapeConfig = {
+    global = filterValidPrometheus cfg.globalConfig;
+    scrape_configs = filterValidPrometheus cfg.scrapeConfigs;
+  };
+
+  filterValidPrometheus = filterAttrsListRecursive (n: v: !(n == "_module" || v == null));
+  filterAttrsListRecursive =
+    pred: x:
+    if isAttrs x then
+      listToAttrs (
+        concatMap (
+          name:
+          let
+            v = x.${name};
+          in
+          if pred name v then [ (nameValuePair name (filterAttrsListRecursive pred v)) ] else [ ]
+        ) (attrNames x)
+      )
+    else if isList x then
+      map (filterAttrsListRecursive pred) x
+    else
+      x;
+in
+{
+  options.services.victoriametrics = {
+    enable = mkEnableOption "VictoriaMetrics, a time series database, long-term remote storage for victoriametrics";
+    package = mkPackageOption pkgs "victoriametrics" { };
+
+    port = mkOption {
+      type = types.port;
+      default = 8428;
+      description = ''
+        Port to listen on.
+      '';
+    };
+
+    listenAddress = mkOption {
+      type = types.str;
+      default = "0.0.0.0";
```

Comment:
In order to be as consistent with Prometheus as possible, I have separated the `port` from the `listenAddress`. This is a breaking change, and I can revert this change if there are any issues.

Reply:
To me this seems a little over the top, I think it's perfectly fine to stick to the victoriametrics way of configuring things here.